### ISSUE TITLE
Concerns regarding lack of separation between indexing and storage

### ISSUE THREAD

**Issue by Ulriksen** [dotnet/orleans#5432]
During our work on the CosmosDB provider we've found some concerns regarding the implementation of indexing and the lack of separation from the storage provider. 

The tl;dr should probably be something in the lines of "storage and indexing is different, and in most cases will require different implementations backing them", there fore it should not be implemented by the same interface. 

A bit longer version:

1. Indexing is implicit in the write state async call, the storage provider must be aware of how to enable indexing of selected paths or able to extract these from the state. 

According to code comments, CosmosDB is currently the only provider supporting this, and some providers such as Blob or table storage will have a hard time supporting it without companion service. 

Services not able to apply indexing directly on the state (Se also 3. on CosmosDB) will have to extract the indexed fields from state them self and store them properly. This seems to require a bit of duplicated and possible complex code in each provider, which could be avoided using the defined methods DirectApplyIndexUpdate and DirectApplyIndexUpdateBatch defined in IIndexInterface and then call to an interface implemented by a (storage- or) Index-Provider. 

The index read query does not read state data, it only returns GrainReference, thus the index besides searching has no use of the state it self. 

2. Reading from index is done via dynamic without a defined interface, although commented with "TODO: define IOrleansIndexingStorageProvider?"

If one are to implement an IndexingStorageProvider it would make even more sense to implement write this way as well.

3. The current implementation in CosmosDB effectively limits a grain type to at most 10GB of data as GrainType is used as partition key. 

The current CosmosDB implementatione queries using stored procedures, which cannot be used in cross partition queries. As partitioning and querying is done by GrainType this limits the amount of data for a single grain type to 10GB. 

A pull request to allow [custom partitioning based on GrainReference](https://github.com/OrleansContrib/Orleans.CosmosDB/pull/26) solves this propblem for general state storage, but does not address indexing. Indexing can still work with this pull request if the CosmosDB provided is change from a stored procedure query to a SQL query  `var query = 'SELECT * FROM c WHERE c.GrainType = "' + grainType + '" AND c.State.' + indexedField + ' = ' + key;`  with feed options set to allow cross partition queries. This will come at a potentially significant cost in terms of request charge on the CosmosDB, and it may be that it will be better to store index data for CosmosDB separately, with only the indexed data and grain reference, partitioned by the grain type while letting the state it self follow custom partition key logic. Implementation of a separate index storage in CosmosDB based on calls from DirectApplyIndexUpdate seems doable.

**Comment by sergeybykov** [comment]
@Ulriksen Should this issue be rather opened in https://github.com/OrleansContrib/OrleansV2.Fork.Indexing? Or do you see a more general problem here? I'm trying to wrap my head around it.

**Comment by Ulriksen** [comment]
Yes, if one could. There's no issues in that repository as it is a fork of this one. 

This is not a general problem.

**Comment by sergeybykov** [comment]
Good point. I missed that. Thanks.

**Comment by philbe** [comment]
@Ulriksen. @TedHartMS and I like your suggestion. But unfortunately, we won't have time to incorporate it in our current round of work, which will be finishing up in the next few weeks.
